lib/eventHubSender.ts

```diff
@@ -0,0 +1,373 @@
+      }
+      if (messages[0].properties) {
+        batchMessage.properties = messages[0].properties;
+      }
+      // Finally encode the envelope (batch message).
+      const encodedBatchMessage = rhea.message.encode(batchMessage);
+      debug(`[${this._context.connectionId}] Sender "${this.name}", ` +
+        `sending encoded batch message.`, encodedBatchMessage);
+      return await this._trySend(encodedBatchMessage, undefined, 0x80013700);
+    } catch (err) {
+      debug("An error occurred while sending the batch message %O", err);
+      throw err;
+    }
+  }
+
+  /**
+   * "Unlink" this sender, closing the link and resolving when that operation is complete.
+   * Leaves the underlying connection/session open.
+   * @method close
+   * @return {Promise<void>} Promise<void>
+   */
+  async close(): Promise<void> {
+    if (this._sender) {
+      try {
+        await rheaPromise.closeSender(this._sender);
+        this.removeAllListeners();
+        delete this._context.senders[this.name!];
+        debug(`Deleted the sender "${this.name!}" from the client cache.`);
+        this._sender = undefined;
+        this._session = undefined;
+        clearTimeout(this._tokenRenewalTimer as NodeJS.Timer);
+        debug(`[${this._context.connectionId}] Sender "${this.name}" closed.`);
+      } catch (err) {
+        debug("An error occurred while closing the sender %O", err);
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Tries to send the message to EventHub if there is enough credit to send them
+   * and the circular buffer has available space to settle the message after sending them.
+   *
+   * We have implemented a synchronous send over here. We shall be waiting for the message
+   * to be accepted or rejected and accordingly resolve or reject the promise.
+   *
+   * @param message The message to be sent to EventHub.
+   * @return {Promise<rheaPromise.Delivery>} Promise<rheaPromise.Delivery>
+   */
+  private _trySend(message: AmqpMessage, tag?: any, format?: number): Promise<rheaPromise.Delivery> {
+    return new Promise((resolve, reject) => {
+      debug(`[${this._context.connectionId}] Sender "${this.name}", credit: ${this._sender.credit}, ` +
+        `available: ${this._sender.session.outgoing.available()}.`);
+      if (this._sender.sendable()) {
+        debug(`[${this._context.connectionId}] Sender "${this.name}", sending message: \n`, message);
+        let onRejected: Func<rheaPromise.Context, void>;
+        let onReleased: Func<rheaPromise.Context, void>;
+        let onModified: Func<rheaPromise.Context, void>;
+        let onAccepted: Func<rheaPromise.Context, void>;
+        const removeListeners = (): void => {
+          this._sender.removeListener("rejected", onRejected);
+          this._sender.removeListener("accepted", onAccepted);
+          this._sender.removeListener("released", onReleased);
+          this._sender.removeListener("modified", onModified);
```

Comment:
>modified [](start = 39, length = 8)

no "received" event for multi-frame messages?

Reply:
@pierreca - Interesting never thought about that. So when the sender gets the received event, it should resolve the promise, correct? 

And btw, do I need to do something special in the sdk for sending multiframe messages or is it something that rhea will take care of automatically and in the sdk I should simply attach an event handler for received event and resolve the promise if that happens?